### PR TITLE
feat(api): handle responses for the patient discovery

### PR DESCRIPTION
**Pull request by Orta21**
Ref. https://github.com/metriport/metriport-internal/issues/1333

### Dependencies

- Upstream: none
- Downstream: none

### Description

Add the internal endpoint to handle the patient discovery response.

### Testing

- Local
   - [x] Sent a patient discovery payload (no match) via postman and made sure it stored correctly
   - [x] Sent a patient discovery payload (with match) via postman and made sure it stored correctly
- Staging
   - [ ] Sent a patient discovery payload (no match) via postman and made sure it stored correctly
   - [ ] Sent a patient discovery payload (with match) via postman and made sure it stored correctly
      
- Production
   - [ ] This shouldnt affect CX data so ill just send a payload to the internal endpoint and see if it gets stored. Then i will manually delete

### Release Plan

This is the release plan for production

- :warning: This contains a DB migration
- [x] publish alpha packages
- [ ] publish prod packages

- [ ] Merge this

### CODE REVIEW

File: packages/api/src/external/patient-external.ts [hunk 1]
```diff
@@ -1,11 +1,12 @@
...
-  };
-}
+// export function getLinkStatusAcrossHIEs(
+//   externalData: PatientExternalData | undefined
+// ): LinkStatusAcrossHIEs {
+//   return {
+//     COMMONWELL: getLinkStatusCW(externalData),
+//     CAREQUALITY: getLinkStatusCQ(externalData),
+//   };
+// }
```
Comment (Orta21) [inline packages/api/src/external/patient-external.ts:12]: This function isnt being used anymore should i remove? It throws error no that ive added CQ

File: packages/api/src/command/medical/patient/create-patient.ts [hunk 2]
```diff
@@ -42,5 +44,12 @@ export const createPatient = async (patient: PatientCreateCmd): Promise<Patient>
...
   cwCommands.patient.create(newPatient, facilityId).catch(processAsyncError(`cw.patient.create`));
 
+  // TODO: NEED CQ PATIENT DIRECTORY SEARCH
+  // const organization = await getOrganizationOrFail({ cxId, id: facilityId });
+
+  // cqCommands.patient
+  //   .discover(newPatient, organization, facilityId)
+  //   .catch(processAsyncError(`cq.patient.create`));
+
   return newPatient;
```
Comment (Orta21) [inline packages/api/src/command/medical/patient/create-patient.ts:54]: This is still a work in progress as it requires the cq directory patient facility locator logic. Also we dont want this logic going into the cloud until the ihe gateway infra is ready. 

This was more for me to wrap my head around the whole flow

File: packages/api/src/external/carequality/api.ts [hunk 3]
```diff
@@ -13,3 +14,20 @@ export function makeCarequalityAPI(apiKey?: string): Carequality | undefined {
...
+/**
+ * Creates a new instance of the IHE Gateway client.
+ * @returns IHE Gateway client.
+ */
+export function makeIheGatewayAPI(): IHEGateway {
+  // TODO ADD IHE GATEWAY MOCK
+  // if (Config.isSandbox()) {
+  //   return;
+  // }
+
```
Comment (Orta21) [inline packages/api/src/external/carequality/api.ts:27]: When we start to handle the IHE Gateway in the cloud then we can add worry about adding this

File: packages/api/src/external/patient-external.ts (hunk 1 above)
Comment (Orta21) [reply packages/api/src/external/patient-external.ts:12]: The question for this and the other comments is whether we want to keep this logic in the PR as comments as it wont run anything or just remove it for now (ill add it to some other branch)

File: packages/api/src/external/carequality/patient.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+      orgOid: organization.oid,
+    });
+
+    iheGateway.startPatientDiscovery(iheGatewayRequest);
+  } catch (err) {
+    console.error(
+      `Failure while starting patient discovery for ${patient.id} @ IHE Gateway: `,
+      err
+    );
+    capture.error(err, {
```
Comment (leite08) [inline packages/api/src/external/carequality/patient.ts:40]: Shouldn't we use the same approach of moving to `capture.message(...)` instead of "error" so we can have a use case oriented message, adding `error` as an extra an `logLevel` as `error`?

File: packages/api/src/external/carequality/patient.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+      extra: {
+        facilityId,
+        patientId: patient.id,
+        context: createContext,
+      },
+    });
+    throw err;
```
Comment (leite08) [inline packages/api/src/external/carequality/patient.ts:47]: Since we're doing error handling here, maybe we can remove the one when we call this function?

File: packages/api/src/external/carequality/patient.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+  if (patientMatch) {
+    await setCarequalityId({
+      patientId,
+      cxId,
+      carequalityPatientId: xcpdPatientId?.id ?? "",
+      carequalityPatientSystemId: xcpdPatientId?.system ?? "",
+    });
+  } else {
+    // WE SHOULD EXPECT THIS TO HAPPEN SO SHOULD WE ADD SENTRY MESSAGE?
+    // DONT WANT TO BOMBARD ALERTS CHANNEL
```
Comment (leite08) [inline packages/api/src/external/carequality/patient.ts:65]: Unless there's something for us to do, I think a console is fine (might even be too much, we also don't want to make logs unusable.

File: packages/api/src/external/patient-external.ts
Comment (leite08) [inline packages/api/src/external/patient-external.ts:1]: This seems like its not being used, so maybe we can just remove it?

File: packages/api/src/routes/medical/internal-cq.ts [hunk 7]
```diff
@@ -81,7 +83,7 @@ router.post(
 
 /**
  * GET /internal/carequality/directory/nearby-organizations
- 
+
```
Comment (leite08) [inline packages/api/src/routes/medical/internal-cq.ts:86]: super nit: remove this empty line

File: packages/api/src/routes/medical/internal-cq.ts [hunk 8]
```diff
@@ -104,4 +106,22 @@ router.get(
...
 
+// BELOW ARE THE ROUTES PERTAINING TO THE IHE-GATEWAY
+
+/**
+ * POST /internal/carequality/patient-discovery
+ *
+ * Receives a Patient Discovery request from the IHE Gateway
+ */
+router.post(
+  "/patient-discovery",
```
Comment (leite08) [inline packages/api/src/routes/medical/internal-cq.ts:117]: ditto: business name, "create"?

File: packages/api/src/routes/medical/internal-cq.ts [hunk 9]
```diff
@@ -104,4 +106,22 @@ router.get(
...
+  asyncHandler(async (req: Request, res: Response) => {
+    const patientDiscovery = patientDiscoveryResponseSchema.parse(req.body);
+
+    await handleDiscoverResponse(patientDiscovery);
```
Comment (leite08) [inline packages/api/src/routes/medical/internal-cq.ts:121]: This endpoint is called by Mirth, right? Do we need to await here, or can we just trigger the Promise and return OK so Mirth doesn't have to wait for it?

File: packages/api/src/external/carequality/api.ts (hunk 3 above)
Comment (Orta21) [reply packages/api/src/external/carequality/api.ts:25]: Well because it wont only be disabling it. I assumed that we would want to mock it the same way we do with CW unless for sandbox having the mock payloads just from CW is enough for our cx's.

File: packages/api/src/command/medical/patient/create-patient.ts [hunk 10]
```diff
@@ -42,5 +44,12 @@ export const createPatient = async (patient: PatientCreateCmd): Promise<Patient>
   // Intentionally asynchronous - it takes too long to perform
   cwCommands.patient.create(newPatient, facilityId).catch(processAsyncError(`cw.patient.create`));
 
+  // TODO: NEED CQ PATIENT DIRECTORY SEARCH
+  // const organization = await getOrganizationOrFail({ cxId, id: facilityId });
```
[the remark this answers is not in the document]
Comment (Orta21) [reply packages/api/src/command/medical/patient/create-patient.ts:48]: The facility your right is unnecessary as for the location i figured since this was logic specific to the domain we would use it here and not on the cq level

File: packages/api/src/external/carequality/api.ts [hunk 11]
```diff
@@ -11,3 +12,20 @@ export function makeCarequalityAPI(apiKey?: string): Carequality | undefined {
...
+  if (Config.isDev() || Config.isStaging()) {
+    return new IHEGateway(APIMode.dev, {});
```
[the remark this answers is not in the document]
Comment (Orta21) [reply packages/api/src/external/carequality/api.ts:27]: Update the params and missed updating it here. Fixed

File: packages/api/src/external/carequality/patient-conversion.ts [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+    cxId: cxId,
+    timestamp: dayjs().toISOString(),
+    xcpdGateways: xcpdGateways,
+    principalCareProviderIds: [facilityNPI],
+    samlAttributes: {
+      subjectId: user,
+      subjectRole: {
+        code: "106331006",
+        display: "Administrative AND/OR managerial worker",
+      },
```
[the remark this answers is not in the document]
Comment (Orta21) [reply packages/api/src/external/carequality/patient-conversion.ts:35]: Yes this is required and its not something we currently ask from the cx's so i figured this role would work hardcoded

File: packages/api/src/external/carequality/patient.ts [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+import { Patient } from "../../domain/medical/patient";
+import { capture } from "../../shared/notifications";
+import { makeIheGatewayAPI } from "./api";
+import { toFHIR } from "../fhir/patient";
+import { patientToIheGateway } from "./patient-conversion";
+import { setCarequalityId } from "./patient-external-data";
+
+const createContext = "cq.patient.create";
+
+export async function discover(
```
[the remark this answers is not in the document]
Comment (Orta21) [reply packages/api/src/external/carequality/patient.ts:12]: Discuss first comment and then will address

File: packages/api/src/external/carequality/patient.ts (hunk 6 above)
Comment (Orta21) [reply packages/api/src/external/carequality/patient.ts:65]: Yea i was thinking at least a log for debugging purposes. Lets say a patient has bad coverage we can look at the logs and see where they werent able to get a match

File: packages/api/src/routes/medical/internal-cq.ts (hunk 8 above)
Comment (Orta21) [reply packages/api/src/routes/medical/internal-cq.ts:117]: Same once comment is addressed

File: packages/api/src/routes/medical/internal-cq.ts (hunk 9 above)
Comment (Orta21) [reply packages/api/src/routes/medical/internal-cq.ts:121]: Yea thats a good point its triggered by Mirth and its not like it expects anything

File: packages/api/src/command/medical/patient/create-patient.ts (hunk 2 above)
Comment (Orta21) [reply packages/api/src/command/medical/patient/create-patient.ts:54]: This will probably be better to talk about in engineering

Comment (Goncharo) [reply packages/api/src/command/medical/patient/create-patient.ts:54]: Regardless of what we do with FF, let's tie this TODO into a specific issue? ie "Add CQ XCPD on Patient Create"

For the FF, I've added it in the Go Live checklist https://github.com/metriport/metriport-internal/issues/1350

File: packages/api/src/external/carequality/api.ts [hunk 14]
```diff
@@ -11,3 +12,20 @@ export function makeCarequalityAPI(apiKey?: string): Carequality | undefined {
...
+ * @returns IHE Gateway client.
+ */
+export function makeIheGatewayAPI(): IHEGateway {
+  // TODO ADD IHE GATEWAY MOCK
+  // if (Config.isSandbox()) {
+  //   return;
+  // }
+
+  if (Config.isDev() || Config.isStaging()) {
+    return new IHEGateway(APIMode.dev);
```
Comment (Goncharo) [inline packages/api/src/external/carequality/api.ts:27]: dev will point to a different endpoint vs staging, right?

dev is local Mirth, staging will be on AWS Mirth

File: packages/api/src/external/carequality/api.ts (hunk 3 above)
Comment (Goncharo) [reply packages/api/src/external/carequality/api.ts:25]: I guess Raf's point is why we don't sure add the piping for the mock now, so we don't have to do it later? 

Like https://github.com/metriport/metriport/blob/baf47279dada489d0d0ced37ef5dfd8cbb2710c5/packages/api/src/external/commonwell/mock.ts#L23

File: packages/carequality-sdk/src/models/purpose-of-use.ts [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+  LAW = "LAW",
+  DECEASED = "DECEASED",
+  DONATION = "DONATION",
+  RESEARCH = "RESEARCH",
+  THREAT = "THREAT",
+  GOVERNMENT = "GOVERNMENT",
+  WORKERSCOMP = "WORKERSCOMP",
+  COVERAGE = "COVERAGE",
+  REQUEST = "REQUEST",
+}
```
Comment (Goncharo) [inline packages/carequality-sdk/src/models/purpose-of-use.ts:30]: https://metriport.slack.com/archives/C065BLRBUDQ/p1701377525668049?thread_ts=1701287906.551129&cid=C065BLRBUDQ

File: packages/api/src/external/carequality/patient-conversion.ts [hunk 16]
```diff
@@ -0,0 +1,43 @@
+import { PatientDiscoveryRequest, XCPDGateways } from "@metriport/ihe-gateway-sdk";
+import { PurposeOfUse } from "@metriport/carequality-sdk";
+import { Patient as FHIRPatient } from "@medplum/fhirtypes";
+import { uuidv7 } from "@metriport/core/util/uuid-v7";
+import dayjs from "dayjs";
+
+export function patientToIheGateway({
```
Comment (Goncharo) [inline packages/api/src/external/carequality/patient-conversion.ts:7]: why don't we call this what it is? ie `createPatientDiscoveryRequest` or something

File: packages/api/src/external/carequality/patient-conversion.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (Goncharo) [reply packages/api/src/external/carequality/patient-conversion.ts:35]: https://metriport.slack.com/archives/C065BLRBUDQ/p1701374311916959?thread_ts=1701287906.551129&cid=C065BLRBUDQ

"we can hardcode it in the sense that we still define the type with all possible options, and just use one of them for now - this way we'll be in a good place for future changes"

let's define a type for this based on https://hl7.org/fhir/R4/valueset-practitioner-role.html - like we do with PurposeOfUse

File: packages/api/src/external/carequality/patient-external-data.ts [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+}): Promise<Patient> => {
+  return executeOnDBTx(PatientModel.prototype, async transaction => {
+    const updatedPatient = await getPatientOrFail({
+      id: patientId,
+      cxId,
+      lock: true,
+      transaction,
+    });
+
+    const updatedData = cloneDeep(updatedPatient.data);
```
[the remark this answers is not in the document]
Comment (Goncharo) [reply packages/api/src/external/carequality/patient-external-data.ts:45]: can we raise a ticket to try this method in both places, then thoroughly test it, and prio accordingly instead?

we've had so much pain with these "sequelize isn't updating the data" scenarios that I don't have confidence this will work 100% of the time in every edge case

File: packages/api/src/routes/medical/internal-cq.ts (hunk 8 above)
Comment (Goncharo) [reply packages/api/src/routes/medical/internal-cq.ts:116]: why is "create" the business name? we're not creating a patient here

patient discovery (PD) is appropriate here - and thought we were on the same page based on this https://metriport.slack.com/archives/C065BLRBUDQ/p1701200344160359?thread_ts=1701037512.758309&cid=C065BLRBUDQ

**Review by Goncharo — CHANGES_REQUESTED** [review]

File: packages/api/src/command/medical/patient/create-patient.ts (hunk 2 above)
Comment (Orta21) [reply packages/api/src/command/medical/patient/create-patient.ts:54]: Should i create that ticket and then move those changes into a new branch so we dont lose them. That way i can clean this up

File: packages/api/src/external/carequality/patient-external-data.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (Orta21) [reply packages/api/src/external/carequality/patient-external-data.ts:46]: https://github.com/metriport/metriport-internal/issues/1353

File: packages/api/src/external/index.ts [hunk 18]
```diff
@@ -12,4 +13,5 @@ export const HL7OID = "2.16.840.1.113883";
 
 export const MedicalDataSourceOid: Record<MedicalDataSource, string> = {
   [MedicalDataSource.COMMONWELL]: `${HL7OID}.3.3330`,
+  [MedicalDataSource.CAREQUALITY]: `${HL7OID}.3.3330`,
```
Comment (Goncharo) [inline packages/api/src/external/index.ts:16]: this is CW http://oid-info.com/get/2.16.840.1.113883.3.3330

we need to get the one for CQ if it exists

Comment (Orta21) [reply packages/api/src/external/index.ts:16]: Doesnt seem like they do will remove

File: packages/api/src/external/carequality/patient.ts (hunk 6 above)
Comment (leite08) [reply packages/api/src/external/carequality/patient.ts]: I like that, but how many entries are we expecting to get on the logs? If its 10, that's fine, if its +1,000 that's likely too much (not saying we shouldn't keep track of it, just saying logs wouldn't be a good solution in this case - maybe a DynamoDB table?).

File: packages/api/src/routes/medical/internal-cq.ts (hunk 8 above)
Comment (leite08) [reply packages/api/src/routes/medical/internal-cq.ts:116]: Yeah, I agree "create" it's not the best name since we're actually searching for patients - in which case, "discovery" seems the most appropriate one for a few reasons.

> thought we were on the same page

Yeah, I thought we were. What does you make you think we're not, just because I suggested the wrong name?

From my understanding, we're giving preference to business operation names instead of tech/protocol oriented (i.e., Doc Query vs ITI-48 - if I got the number right).

Now that we're here, I'm thinking that `POST /patient-discovery` doesn't do what it "sounds" like. One needs explanation that this endpoint doesn't actually do PD, but it's the response of a PD request.

Maybe append "/response"?

`POST /patient-discovery/response`

File: packages/api/src/sequelize/migrations/2023-12-12_00_create-patient-discovery.ts [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+export const up: Migration = async ({ context: queryInterface }) => {
+  await queryInterface.sequelize.transaction(async transaction => {
+    await shared.createTable(
+      queryInterface,
+      tableName,
+      {
+        id: {
+          type: DataTypes.STRING,
+          primaryKey: true,
+        },
```
Comment (leite08) [inline packages/api/src/sequelize/migrations/2023-12-12_00_create-patient-discovery.ts:17]: Do we need ID on this table? Asking b/c it generates an index and I'm not sure we need an ID. Or could create the ID as UUID and not mark it as PK (so no index).

File: packages/api/src/sequelize/migrations/2023-12-12_00_create-patient-discovery.ts [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+        status: {
+          type: DataTypes.STRING,
+          field: "status",
+        },
+        data: {
+          type: DataTypes.JSONB,
+        },
+      },
+      { transaction, addVersion: true }
```
Comment (leite08) [inline packages/api/src/sequelize/migrations/2023-12-12_00_create-patient-discovery.ts:34]: Likely don't need version on this table, right?

Actually, might consider whether we even need `updatedAt` - if not, either don't use `shared.createTable()` or update it to allow `updatedAt` to be parameterized like `version`.

File: packages/api/src/sequelize/migrations/2023-12-12_00_create-patient-discovery.ts [hunk 21]
```diff
@@ -0,0 +1,43 @@
+import { DataTypes } from "sequelize";
+import type { Migration } from "..";
+import * as shared from "../migrations-shared";
+
+const tableName = "patient_discovery_result";
```
Comment (leite08) [inline packages/api/src/sequelize/migrations/2023-12-12_00_create-patient-discovery.ts:5]: Do we know how we'll query this table? We can add indexes later, but if we know that we'll query for created+status or just created, could create the index out of the bat and maybe save a DB migration later.

NABD otherwise.

**Review by leite08 — APPROVED** [review]
LGTM, just added a few considerations about the table considering we'll store a lot of data there. But none block the release.

I'd also make sure to have a ticket/item on the scope of this initiative to deal with old data, like a sweeper lambda to clean up data older than X days.

File: packages/api/src/sequelize/migrations/2023-12-12_00_create-patient-discovery.ts (hunk 19 above)
Comment (Orta21) [reply packages/api/src/sequelize/migrations/2023-12-12_00_create-patient-discovery.ts:17]: I believe we still want the id but ill remove the primarykey

File: packages/api/src/sequelize/migrations/2023-12-12_00_create-patient-discovery.ts (hunk 20 above)
Comment (Orta21) [reply packages/api/src/sequelize/migrations/2023-12-12_00_create-patient-discovery.ts:34]: Good point will update

File: packages/api/src/sequelize/migrations/2023-12-12_00_create-patient-discovery.ts (hunk 21 above)
Comment (Orta21) [reply packages/api/src/sequelize/migrations/2023-12-12_00_create-patient-discovery.ts:5]: I see it as a query for all results in a certain time period for a given patient

File: packages/ihe-gateway-sdk/src/models/shared.ts [hunk 22]
```diff
@@ -64,7 +64,8 @@ export const baseResponseSchema = z.object({
   cxId: z.string(),
   timestamp: z.string(),
   responseTimestamp: z.string(),
-  xcpdPatientId: z.object({ id: z.string(), system: z.string() }).nullable(),
+  xcpdPatientId: z.object({ id: z.string(), system: z.string() }).optional(),
+  gateway: z.object({ oid: z.string(), url: z.string() }),
   patientId: z.string(),
-  operationOutcome: operationOutcome.nullable(),
+  operationOutcome: operationOutcome.optional(),
```
Comment (leite08) [inline packages/ihe-gateway-sdk/src/models/shared.ts:70]: AFAIK we moved from `optional()` and `nullable()` to `nullish()`, which include both null and undefined. Not applicable here and other schemas in this PR?

File: packages/commonwell-sdk/src/common/make-jwt.ts [hunk 23]
```diff
@@ -1,5 +1,5 @@
 import * as jwt from "jsonwebtoken";
-import { PurposeOfUse } from "../models/purpose-of-use";
+import { PurposeOfUse } from "@metriport/shared";
```
Comment (leite08) [inline packages/commonwell-sdk/src/common/make-jwt.ts:2]: I don't think we can do this. Shared AFAIK is not published and we don't want to publish it, so we can keep it streamlined the development without publishing packages, no?

File: packages/commonwell-sdk/src/common/util.ts [hunk 24]
```diff
@@ -3,7 +3,7 @@ import { StrongId } from "../models/identifier";
 import { Organization } from "../models/organization";
 import { Patient } from "../models/patient";
 import { Person, PersonSearchResp } from "../models/person";
-import { PurposeOfUse } from "../models/purpose-of-use";
+import { PurposeOfUse } from "@metriport/shared";
```
Comment (leite08) [inline packages/commonwell-sdk/src/common/util.ts:6]: ditto shared on SDKs

File: packages/commonwell-sdk/src/common/make-jwt.ts (hunk 23 above)
Comment (leite08) [reply packages/commonwell-sdk/src/common/make-jwt.ts:2]: Oh, it's [published](https://www.npmjs.com/package/@metriport/shared)! 🤔

Comment (Orta21) [reply packages/commonwell-sdk/src/common/make-jwt.ts:2]: I thought this would reference the local package as it was getting built, if not what would be the best approach to prevent duplicating this?

**Comment by leite08** [comment]
The release plan is missing updates on NPM packages, and also the PR is missing the `alpha` versions

File: packages/api/src/sequelize/migrations/2023-12-12_00_create-patient-discovery.ts (hunk 21 above)
Comment (leite08) [reply packages/api/src/sequelize/migrations/2023-12-12_00_create-patient-discovery.ts:4]: I see we're not setting the ID as PK but also not creating any indexes. If the query you mentioned is the one we're going to use, maybe we define this index now? Or can you ticket both the index and the sweeper logic just so we don't forget as part of the go live, please? Even if it comes a couple weeks afterwards, it's defs something we'll need sooner than later, as getting to 1MM records on this table won't take long.

**Review by Goncharo — APPROVED** [review]
🚀

**Review by leite08 — APPROVED** [review]
